_posts/2021-08-23-using-lra.adoc

```diff
@@ -0,0 +1,490 @@
+----
+$ java -Dquarkus.http.port=8081 -jar target/quarkus-app/quarkus-run.jar &
+[2] 2495369
+----
+
+Since we are still using the same application resource file and external coordinator there is no need to update the config.
+
+Again, start an LRA using curl to send a request to the start method of the first service:
+
+[source,bash]
+----
+$ LRA_URL=$(curl http://localhost:8080/hello/start)
+hello with context http://localhost:50000/lra-coordinator/0_ffffc0a8000e_a355_6113dede_11
+----
+
+and now end it in the second service:
+
+[source,bash]
+----
+$ curl --header "Long-Running-Action: $LRA_URL" http://localhost:8081/hello
+hello with context http://localhost:50000/lra-coordinator/0_ffffc0a8000e_a355_6113dede_11
+completing http://localhost:50000/lra-coordinator/0_ffffc0a8000e_a355_6113dede_11
+completing http://localhost:50000/lra-coordinator/0_ffffc0a8000e_a355_6113dede_11
+Hello RESTEasy
+----
+
+Notice that both microservices indicated that they received the completion callback.
+
+Terminate both java processes (`kill 2495275 2495369`).
+
+==== Optional Step: using the MANDATORY element
+
+Instead of using an existing method to close the LRA you might prefer to write one which expects there to be a context.
+In this case you would want to set the `LRA.Type` element:
+
+[source,java]
+----
+    @GET
+    @Path("/end")
+    @LRA(value = LRA.Type.MANDATORY) // Step 3a: The method MUST be invoked with an LRA
+    @Produces(MediaType.TEXT_PLAIN)
+    public String end(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return lraId.toASCIIString();
+    }
+----
+
+Because the end method is annotated with `@LRA(value = LRA.Type.MANDATORY)`, the context header must be present otherwise the method will return an error response code:
+
+[source,bash]
+----
+$ ./mvnw clean package -DskipTests
+$ java -Dquarkus.http.port=8081 -jar target/quarkus-app/quarkus-run.jar &
+[1] 300189
+$ LRA_URL=$(curl http://localhost:8081/hello/start)
+$ curl -v http://localhost:8081/hello/end
+...
+HTTP/1.1 412 Precondition Failed
+...
+----
+
+whereas providing the LRA context header will work:
+
+[source,bash]
+----
+$ curl --header "Long-Running-Action: $LRA_URL" -I http://localhost:8081/hello/end
+HTTP/1.1 200 OK
+Content-Type: application/octet-stream
+connection: keep-alive
+$ kill 300189
+----
+
+Other https://download.eclipse.org/microprofile/microprofile-lra-1.0/apidocs/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.Type.html[LRA.Type element values] may be useful, depending upon what your application is trying to achieve. For those readers familiar with JTA it is worth remarking that it was loosely modelled on the https://docs.oracle.com/javaee/7/api/javax/transaction/Transactional.TxType.html[Java Transactional TxType annotation].
+
+== Native mode
+
+NOTE: in native mode only external coordinators (i.e. not embedded with the application) are supported (we will provide a coordinator extension in a later release to address this deficiency).
```

Comment:
The text above contains similar remark
(`(NOTE: you cannot use this approach to run coordinators in native mode, a future extension will be provided to support this requirement).`)
isn't this one an duplication?

Reply:
The note concerns running coordinators in native mode and it is important to include it in the native mode section and to also  prefix it with an asciidoc `NOTE:`. Therefore if I remove the initial reference, in the section on building coordinators, then the user may assume that he can produce native coordinators, and only if he reads the later section will he realise that his initial assumption was incorrect. Furthermore, if the user came to the section via a google search then he would potentially miss it altogether).

Secondly, the second note is referring to embedded coordinators whereas the first note is referring to building standalone native coordinators and therefore it is subtly different from the second reference.

Reply:
Ok, I see.